lib/compare.js

```diff
@@ -95,8 +95,45 @@ function errorCompare(err1, err2) {
   );
 }
 
+function equalWithCircular(obj1, obj2, ref1 = [], ref2 = []) {
+  ref1.push(obj1);
+  ref2.push(obj2);
+
+  if (Array.isArray(obj1) !== Array.isArray(obj2)) return false;
+
+  for (const key in obj1) {
+    const val1 = obj1[key];
+    const val2 = obj2[key];
+
+    const index1 = ref1.indexOf(val1);
+    if (index1 === -1) {
+      if (
+        typeof val1 === 'object' &&
+        val1 &&
+        typeof val2 === 'object' &&
+        val2
+      ) {
+        if (!equalWithCircular(val1, val2, [...ref1], [...ref2])) {
+          return false;
+        }
+      } else if (typeof val1 === 'function' && !equalFunction(val1, val2)) {
+        return false;
+      } else if (Number.isNaN(val1) !== Number.isNaN(val2)) {
+        return false;
+      } else if (val1 !== val2) {
+        return false;
+      }
+    } else if (index1 !== ref2.indexOf(obj2[key])) {
+      return false;
+    }
+  }
+
+  return true;
+}
```

Comment:
Consider this, it may have a few bugs but makes this a bit more consistent with current comparison.
```js
function equalWithCircular(obj1, obj2, ref1 = [], ref2 = []) {
  ref1.push(obj1);
  ref2.push(obj2);

  if (obj1 === obj2) return true;
  if (obj1 === null || obj2 === null) return obj1 === obj2;
  if (Array.isArray(obj1) !== Array.isArray(obj2)) return false;

  const keys1 = Object.keys(obj1);
  const keys2 = Object.keys(obj2);
  if (keys1.length !== keys2.length) return false;

  return keys1.every(key => {
    const val1 = obj1[key];
    const val2 = obj2[key];

    const type1 = typeof val1;
    const type2 = typeof val2;

    if (type1 !== type2) return false;

    const index1 = ref1.indexOf(val1);
    if (index1 !== -1) {
      return index1 === ref2.indexOf(val2);
    }
    if (type1 === 'object') {
      return equalWithCircular(val1, val2, [...ref1], [...ref2]);
    }
    if (type1 === 'function') {
      return equalFunction(val1, val2);
    }
    return Number.isNaN(val1) && Number.isNaN(val2);
  });
}
```

I've added a few checks around `Array`s, `null`s and used `Object.keys` as we have it in `strictEqual`.

Object.keys is not ideal, but we should make this function consistent with what we already have.